### ISSUE TITLE
The average function for float16 can overflow

### ISSUE THREAD

**Issue by gwtnb** [chainer/chainer#7681]
The average function for float16 [can overflow](https://github.com/chainer/chainer/blob/6fef53f3f9fcae9de0643d677b89349d58f20cad/chainer/functions/math/average.py#L62) due to the same reason described in #6702.

**Comment by toslunar** [comment]
Using `xp.average` might not a good solution because
```
>>> np.__version__, np.average(np.full((300,), 500, dtype=np.float16))
('1.9.3', inf)
```
```
>>> np.__version__, np.average(np.full((300,), 500, dtype=np.float16))
('1.16.4', 500.0)
```

**Comment by toslunar** [comment]
Using `xp.average` might not be a good solution because
```
>>> np.__version__, np.average(np.full((300,), 500, dtype=np.float16))
('1.9.3', inf)
```
```
>>> np.__version__, np.average(np.full((300,), 500, dtype=np.float16))
('1.16.4', 500.0)
```

**Comment by toslunar** [comment]
It's fixed for numpy>=1.12 and cupy.
